File: ml-benchmark/synduce/numbers/int_nat_twosum.f
Inductive Nat = z Unit | s Nat;
Inductive INat = positive Nat | negative Nat;
Inductive TwoInats = twoinats {INat, INat};

Inductive EInt = zero Unit | sub1 EInt | add1 EInt;
Inductive TwoInts = twoints {EInt, EInt};

nsum = fix (
  \f: Nat -> Int. \n: Nat.
  match n with
    z _ -> 0
  | s m -> + 1 (f m)
  end
);

itoint = \n: INat.
  match n with
    negative m -> - 0 (+ 1 (nsum m))
  | positive m -> nsum m
  end;

two_isum = \n: TwoInats. 
  match n with
    twoinats {n1, n2} -> + (itoint n1) (itoint n2)
  end;

target_eint = fix (
  \f: EInt -> Compress EInt. \e: EInt.
  match e with
    zero _ -> zero unit
  | add1 e' -> add1 (f e')
  | sub1 e' -> sub1 (f e')
  end
);

target_twoints = \x: TwoInts.
  match x with
    twoints {x1, x2} -> twoints {target_eint x1, target_eint x2}
  end;

irepr = fix (
  \f: EInt -> INat. \e: EInt.
  match e with
    zero _ -> positive (z unit)
  | sub1 e' ->
    let res = f e' in
    match res with
      positive (z _) -> negative (z unit)
    | positive (s n) -> positive n
    | negative n -> negative (s n)
    end
  | add1 e' ->
    let res = f e' in
    match res with
      positive n -> positive (s n)
    | negative (z _) -> positive (z unit)
    | negative (s n) -> negative n
    end
  end
);

repr = \x: TwoInts.
  match x with
    twoints {x1, x2} -> twoinats {irepr x1, irepr x2}
  end;

main = \x: TwoInts. two_isum (repr (target_twoints x));

File: ml-benchmark/synduce/numbers/int_nat_twosum.ml
type nat = Z of unit | S of nat
type inat = Positive of nat | Negative of nat
type twoinats = TwoInats of inat * inat

type eint = Zero of unit | Sub1 of eint | Add1 of eint
type twoints = TwoInts of eint * eint

let rec nsum n =
  match n with
  | Z _ -> 0
  | S m ->
      let r = nsum m in
      1 + r

let itoint n =
  match n with
  | Negative m ->
      let s = nsum m in
      0 - (1 + s)
  | Positive m ->
      nsum m

let two_isum n =
  match n with
  | TwoInats (n1, n2) ->
      let a = itoint n1 in
      let b = itoint n2 in
      a + b

let rec target_eint e =
  match e with
  | Zero _ -> Zero ()
  | Add1 e' ->
      let r = target_eint e' in
      Add1 r
  | Sub1 e' ->
      let r = target_eint e' in
      Sub1 r

let target_twoints x =
  match x with
  | TwoInts (x1, x2) ->
      let y1 = target_eint x1 in
      let y2 = target_eint x2 in
      TwoInts (y1, y2)

let rec irepr e =
  match e with
  | Zero _ -> Positive (Z ())
  | Sub1 e' ->
      let res = irepr e' in
      (match res with
       | Positive n1 ->
           (match n1 with
            | Z _ -> Negative (Z ())
            | S n -> Positive n)
       | Negative n ->
           let n' = S n in
           Negative n')
  | Add1 e' ->
      let res = irepr e' in
      (match res with
       | Positive n ->
           let sn = S n in
           Positive sn
       | Negative n1 ->
           (match n1 with
            | Z _ -> Positive (Z ())
            | S n -> Negative n))

let repr x =
  match x with
  | TwoInts (x1, x2) ->
      let y1 = irepr x1 in
      let y2 = irepr x2 in
      TwoInats (y1, y2)

let program x =
  let tx = target_twoints x in
  let r = repr tx in
  two_isum r